src/extension_commands.ml

```diff
@@ -534,6 +556,105 @@ module Copy_type_under_cursor = struct
     command Extension_consts.Commands.copy_type_under_cursor handler
 end
 
+module Construct = struct
+  let extension_name = "Construct"
+
+  let is_valid_text_doc textdoc =
+    match TextDocument.languageId textdoc with
+    | "ocaml" | "ocaml.interface" | "reason" | "ocaml.ocamllex" -> true
+    | _ -> false
+
+  let ocaml_lsp_doesnt_support_construct ocaml_lsp =
+    not (Ocaml_lsp.can_handle_construct ocaml_lsp)
+
+  let get_construct_results position text_editor client =
+    let doc = TextEditor.document text_editor in
+    let uri = TextDocument.uri doc in
+    Custom_requests.(
+      send_request
+        client
+        Construct.request
+        (Construct.make ~uri ~position ~depth:None ~with_values:None ()))
+
+  let rec display_results ~text_editor:_ ~position:_
+      (results : Custom_requests.Construct.response) =
+    let quickPickItems =
+      List.map results.result ~f:(fun res ->
+          (QuickPickItem.create ~label:res (), (res, results.position)))
+    in
+    let quickPickOptions =
+      QuickPickOptions.create ~title:"Construct results" ()
+    in
+    Window.showQuickPickItems
+      ~choices:quickPickItems
+      ~options:quickPickOptions
+      ()
+
+  and process_construct position text_editor client instance =
+    let open Promise.Syntax in
+    let* res = get_construct_results position text_editor client in
+    let* selected_result = display_results ~text_editor ~position res in
+    match selected_result with
+    | Some (value, range) -> (
+      let* value_inserted = insert_to_document text_editor range value in
+      match value_inserted with
+      | true ->
+        let* new_range =
+          Holes_commands.hole_position text_editor client `Prev
```

Comment:
This will go to the last hole if there are multiple ones in the result, which is not very natural. I think you should have ` Holes_commands.hole_position` take a position as parameter and not a text_editor. That way you can call it with a position that corresponds to the beginning of the inserted range:

```
Holes_commands.hole_position (Range.start range) client `Next
```

Reply:
Yes indeed it goes to the last hole before coming to the hole before it

Reply:
`Hole_commands.hole_position` calls ocamllsp to get the list of hole positions, which is why it's receiving `text_editor` and `client`